### ISSUE TITLE
JavaScript Algorithms and Data Structures (Beta) - Certification Projects - Directing campers away from incorrect solution

### ISSUE THREAD

**Issue by gikf** [freeCodeCamp/freeCodeCamp#53836]
**Is your feature request related to a problem? Please describe.**
Another issue often encountered by campers in Certification Projects is relying on code defined in global context, that's executed just once (when preview is loaded).

This is hard for campers, as any manual change to code will reload preview and therefore everything appears to be working fine, when checking it _manually_.

**Describe the solution you'd like**
Ideal would be finding way to direct campers away from such dependency. So far I've thought about two things:
- Additional note in instructions.
- Test targeting such situations (in other words, test that will make hard or impossible to pass when code depending on other globals than it's expected).

**Comment by jdwilkin4** [comment]
> Test targeting such situations (in other words, test that will make hard or impossible to pass when code depending on other globals than it's expected).

I personally vote for this option

**Comment by JoyShaheb** [comment]
i will vote for the last option as well, @jdwilkin4 let;s open it up for contributors ? lets see the situation and merge the best solutions.

**Comment by jdwilkin4** [comment]
I am fine with opening this up for help wanted contribution.
But it would be best to open up separate PR's for each certification project.
It will be cleaner to review that way.

- [ ] Palindrome Checker
- [ ] Roman Numeral converter
- [ ] Cash Register Project
- [ ] Pokemon project
- [ ] Telephone validator Project

**Comment by a2937** [comment]
How would we test to see if the camper uses globals?

**Comment by gikf** [comment]
I was tinkering with tracking, what global variables might have been added by campers. As much as there's few more or less hacky ways, like:

Listing all variables added after certain point (watch out for variables hoisting):
```javascript
const globals = Object.keys(globalThis);  // add at the top of file
console.log(globals.slice(globals.indexOf('globals') + 1));
```

Or even changing specific variables to getter/setter :eyes: :

```javascript
Object.defineProperty(globalThis, 'price', {
  price: 0,
  get() {
    console.log('get', this.price);
    return this.price
  },
  set(value) {
    this.price = value;
    console.log('set', value);
  }
});
```

That's not going to help much with determining, if that global variable is creating issues. Adding constraint for number of used global variables would limit also valid usages.

**Comment by gikf** [comment]
How's this for hint text?
> Your code should work properly when `cid` and `price` are reassigned in the background, after the initial project load.

**Comment by gikf** [comment]
Some more regarding Cash Register.

- It might be good to change `cid` in the seed to lower amounts, or zeros. High initial values in `cid` is something that can make debugging harder for campers, when they make calculations in global scope on the initial values.
- Hard issue to narrow down is when campers are using global for returned change, to which values are only added to. Would it make sense to add test specifically checking result after clicking button twice?
- Tests are actually not checking if result contains other denominations than expected. One way of correcting this could be making sure total change due in result is what's expected.

**Comment by a2937** [comment]
Okay let's see if I am on the right page.

1) Add a test involving insufficient funds near the bottom . 
2) Add a test that clicks the button twice 
3) Check all coin denominations in test involving purchases

**Comment by gikf** [comment]
1. What I specifically would want to do with changing `cid` in seed is making 
> When price is 3.26, the value in the #cash element is 100, cid is [["PENNY", 1.01], ["NICKEL", 2.05], ["DIME", 3.1], ["QUARTER", 4.25], ["ONE", 90], ["FIVE", 55], ["TEN", 20], ["TWENTY", 60], ["ONE HUNDRED", 100]], and the #purchase-btn element is clicked, the value in the #change-due element should be "Status: OPEN TWENTY: $60 TEN: $20 FIVE: $15 ONE: $1 QUARTER: $0.5 DIME: $0.2 PENNY: $0.04"

Test not pass, when _correctly_ testing manually, and when somebody relies on some calculations done in global scope on initial `cid`. Ie. calculates total cash.

2. Yeah. And just to be sure fill again the `#cash` element, in case camper is clearing them after the button click.
3. Yeah, make sure these denominations that should not be used are not in result.

**Comment by a2937** [comment]
@gikf I was hoping to add those tests as part of my pull request.

**Comment by gikf** [comment]
@a2937 sorry, I had that one started even before writing here. But if you want you can include these changes, I don't mind.

**Comment by UtsavKashyap1710** [comment]
It sounds like your feature request is addressing an issue where learners (referred to as "campers") rely on globally defined code in projects, which gets executed only once when the preview loads. This can create confusion, as any manual code changes that reload the preview make it seem like everything is functioning properly, even when it may not be under different conditions or during testing.

Problem Description
When campers rely on globally defined code (e.g., variables, functions in the global scope), it can lead to inaccurate testing results or unexpected behavior. Since the global code only executes once during the initial preview, any modifications or tests run afterward can give the impression that the code is correct when, in fact, it may fail under different execution conditions (such as being triggered multiple times).

Proposed Solution
Additional Notes in Instructions:

Including a specific section in the project instructions that explains the issues with global code dependencies and encourages campers to structure their code in a way that doesn't rely on global context.
Highlight the importance of isolating code logic and ensuring that functions or variables behave as expected when called multiple times or in different contexts.
Example:

"Ensure that your code does not rely on global variables or functions executed at the start. Your solution should work under multiple test scenarios, including reloading or running the code more than once without resetting the environment."

Test Targeting Global Dependencies:

Design tests that specifically catch code depending on global variables or single-time execution. For instance:
Simulate multiple runs: Create test cases that simulate the function or code being run multiple times without reloading, which would break if the camper is relying on a global setup.
Scope isolation tests: Add test cases where the global context is altered or non-existent, ensuring that the camper's solution doesn’t depend on it.
Example:

A test could clear global variables between runs to ensure the solution doesn’t break when certain global states are missing.
Tests could be designed to execute multiple instances of the function (or simulate multiple page loads) to ensure the code runs correctly each time without relying on initial global state.
Benefits
This approach would encourage better coding practices by pushing campers to avoid global dependencies, making their code more robust and modular.
It would help campers understand the importance of test-driven development and the implications of global variables in dynamic environments.
By applying these two strategies, the learning experience could be significantly improved, helping campers develop better coding habits and ensuring their solutions are resilient in various scenarios.

**Comment by ojeytonwilliams** [comment]
Hey @gikf to make this clearer for potential contributors, can we clarify the goal? Are we trying to stop campers from providing solutions that modify values that should be constant?

i.e. in the cash register we don't want user code to mutate `cid`?

**Comment by Samet047** [comment]
I'll take this issue

**Comment by a2937** [comment]
> I'll take this issue

We typically do not assign issues. Instead, we accept the first pull request that comprehensively solves the issue.

Issues labeled with `help wanted` or `first timers only` are open for contributions.

Please make sure you read [our guidelines for contributing](https://contribute.freecodecamp.org/#/). We prioritize contributors following the instructions in our guide. Join us in [our chat room](https://discord.gg/PRyKn3Vbay) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing - our community will be happy to assist you.

**Comment by majestic-owl448** [comment]
I propose closing this, unless there is a reason not to

**Comment by anishlukk123** [comment]
I  honestly dont see a reason not to close this
